cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/batch/source/FileBatchSource.java

```diff
@@ -0,0 +1,201 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.template.etl.batch.source;
+
+import co.cask.cdap.api.annotation.Description;
+import co.cask.cdap.api.annotation.Name;
+import co.cask.cdap.api.annotation.Plugin;
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.cdap.api.dataset.DatasetProperties;
+import co.cask.cdap.api.dataset.lib.KeyValue;
+import co.cask.cdap.api.dataset.lib.KeyValueTable;
+import co.cask.cdap.api.templates.plugins.PluginConfig;
+import co.cask.cdap.template.etl.api.Emitter;
+import co.cask.cdap.template.etl.api.PipelineConfigurer;
+import co.cask.cdap.template.etl.api.batch.BatchSource;
+import co.cask.cdap.template.etl.api.batch.BatchSourceContext;
+import co.cask.cdap.template.etl.common.BatchFileFilter;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.io.LongWritable;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.text.SimpleDateFormat;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * A {@link BatchSource} to use any distributed file system as a Source.
+ * Currently only supports S3.
+ */
+@Plugin(type = "source")
+@Name("S3")
+@Description("Batch source for S3")
+public class FileBatchSource extends BatchSource<LongWritable, Object, StructuredRecord> {
+
+  private KeyValueTable table;
+  private Date prevMinute;
+
+  private static final String REGEX_DESCRIPTION = "Regex to filter out filenames in the path. " +
+    "To use the TimeFilter, input \"timefilter\". The TimeFilter assumes that it " +
+    "are reading in files with the S3 log naming convention of YYYY-MM-DD-HH-mm-SS-Tag. The TimeFilter " +
+    "reads in files from the previous hour if the timeTable field is left blank. So if it's currently " +
+    "2015-06-16-15 (June 16th 2015, 3pm), it will read in files that contain 2015-06-16-14 in the filename. " +
+    "If the field timeTable is present, then it will read files in that haven't been read yet.";
+  private static final String ACCESS_ID_DESCRIPTION = "AccessID of the S3 source.";
+  private static final String ACCESS_KEY_DESCRIPTION = "AccessKey of the S3 source.";
+  private static final String PATH_DESCRIPTION = "Path to file(s) to be read. If a directory is specified, " +
+    "terminate the path name with a \'/\'";
+  private static final String TABLE_DESCRIPTION = "Name of the Table that keeps track of the last time files " +
+    "were read in.";
+  private static final String INPUT_FORMAT_CLASS_DESCRIPTION = "Name of the input format class. Defaults to " +
```

Comment:
Mention in the description that this class should be a subclass of ``FileInputFormat``